### ISSUE TITLE
setup a 64 core jupyter system

### ISSUE THREAD

**Issue by robnagler** [radiasoft/devops#74]
This could be a multi-node system. Will be used by @cchall for synergia so works well with MPI.

1gb/core is minimum.

1TB disk. Does not need to be backed up.

**Comment by robnagler** [comment]
Scaleway failed due to NATted nets. Synergia performance on Linode 4 cores x 16 nodes was awful so went with 4 independent 20 core nodes. Linode seems to scale cores at 75% with SRW and more poorly with Synergia.

Blades still are the fastest.

**Comment by robnagler** [comment]
Ran some tests on the blade, and seeing 75% scaling there, too. SRW uses MPI rank 0 as a control so I exclude this node from the seconds per particle (s/p) calculations. You can see in both cases as the number of cores goes up, the efficiency drops by 75% after about 5 cores. From 1 to 5 cores, 100% efficiency.

Linode 20c (Linode 80G) on SSD (docker root):
1c 5p 134s 27s/p
20c 57p 131s 43s/p
10c 18p 65s 32s/p
5c 20p 128s 26s/p
11c 20p 67s 34s/p
17c 32p 86s 43s/p
9c 32p 115s 29s/p

Dell em620 blade on SSD (docker root):
1c 5p 103s 21s/p
5c 20p 105s 21s/p
11c 20p 58s 29s/p
16c 30p 61s 30s/p

Image: https://registry.hub.docker.com/radiasoft/beamsim-jupyter:20170315.041636

**Comment by bruhwile** [comment]
Great results - thanks!
I think it would be good to follow some standard protocols for testing
parallel performance.

"Weak scaling" is used to understand how the computing and communication
system is performing, under the assumption that there are no fundamental
problems with the software:
1) minimize any initial reading of files
2) turn off all file writing that isn't absolutely essential
3) keep the work per processor constant --> problem size grows linearly
with number of cores
4) make sure the work per processor is large enough that serial overhead
doesn't obscure the results
5) use identical runs to compare different hardware configurations

For SRW, item 4) means (I think) that at least 10 macro-particles per core
should be used.

Once the weak scaling is understood, "strong scaling" is used to understand
limitations of a code's MPI implementation, for a specific hardware
configuration, installed MPI, etc.
1) minimize any initial reading of files
2) turn off all file writing that isn't absolutely essential
3) keep the problem size fixed

For SRW, I would take 10 x max_num_procs as the fixed number of
macro-particles.
The pain here is that the single-processor simulation will take a long time.

Finally, there is "practical scaling" for lack of a better term.
The idea is to repeat the strong scaling exercise with typical I/O enabled.

In summary, weak scaling helps you decide whether a particular computing
system is acceptable as compared to other systems. The strong scaling lets
users know how many processors they can effectively use (best case
scenario) on a specific system. "Practical scaling" provides more realistic
guidance for users as to how many processors they can really use
effectively. Looking at results for all three types of scaling can help to
diagnose the limiting factor(s).

D.

**Comment by robnagler** [comment]
I got a bit more scientific and wrote [srw-timer.sh](https://github.com/radiasoft/devops/blob/7033d783bcb59dd4302335b8083697eaa65ebe6a/bin/srw-timer.sh), which makes it easier to do runs. The `s/p/w` result discounts the one core dedicated to coordination. That's maybe not "fair" for small

SRW is efficient so once you get to three macro particles, you can get a good sense of scaling performance. I/O is efficient, too, so I would say the strong and weak cases are pretty much the same.

Linode does not scale linearly, which is strange. The Blade and MacBook VirtualBox scale linearly at three particles.

### Dell m620 Blade

#### apa20 16 cores, E5-2650 v2 @ 2.60GHz, 128gb DDR3

1c **1p** 30s 30s/p/w
5c 4p 33s 33s/p/w
9c 8p 37s 37s/p/w
16c 15p 41s 41s/p/w

1c **2p** 48s 24s/p/w
5c 8p 50s 25s/p/w
9c 16p 56s 28s/p/w
16c 30p 61s 31s/p/w

1c **3p** 67s 22s/p/w
5c 12p 70s 23s/p/w
9c 24p 76s 25s/p/w
16c 45p 80s 27s/p/w

1c **4p** 83s 21s/p/w
5c 16p 88s 22s/p/w
9c 32p 98s 25s/p/w
16c 60p 100s 25s/p/w

1c **5p** 104s 21s/p/w
5c 20p 105s 21s/p/w
9c 40p 117s 23s/p/w
16c 75p 119s 24s/p/w

### Linode 20 cores, 80gb

#### dfw10 E5-2680 v2 @ 2.80GHz

1c **5p** 128s 26s/p/w
5c 20p 128s 26s/p/w
10c 45p 136s 27s/p/w
20c 95p 194s 39s/p/w

1c **5p** 131s 26s/p/w
5c 20p 127s 25s/p/w
10c 45p 135s 27s/p/w
20c 95p 187s 37s/p/w

#### dfw11 E5-2680 v3 @ 2.50GHz

1c **5p** 113s 23s/p/w
5c 20p 128s 26s/p/w
10c 45p 175s 35s/p/w
20c 95p 233s 47s/p/w

1c **5p** 120s 24s/p/w
5c 20p 144s 29s/p/w
10c 45p 175s 35s/p/w
20c 95p 245s 49s/p/w

#### dfw12 E5-2680 v3 @ 2.50GHz

1c **5p** 118s 24s/p/w
5c 20p 127s 25s/p/w
10c 45p 161s 32s/p/w
20c 95p 212s 42s/p/w

1c **5p** 129s 26s/p/w
5c 20p 125s 25s/p/w
10c 45p 170s 34s/p/w
20c 95p 232s 46s/p/w

#### dfw13 E5-2680 v2 @ 2.80GHz

1c **5p** 130s 26s/p/w
5c 20p 133s 26s/p/w
9c 40p 157s 31s/p/w
19c 90p 216s 43s/p/w

1c **5p** 130s 26s/p/w
5c 20p 127s 25s/p/w
10c 45p 136s 27s/p/w
20c 95p 192s 38s/p/w

1c **5p** 130s 26s/p/w
5c 20p 128s 26s/p/w
10c 45p 135s 27s/p/w
20c 95p 186s 37s/p/w

#### dfw13 nfs
1c **2p** 59s 29s/p/w
5c 8p 62s 31s/p/w
9c 16p 64s 32s/p/w
19c 36p 89s 44s/p/w

1c **5p** 129s 25s/p/w
5c 20p 134s 26s/p/w
9c 40p 155s 31s/p/w
19c 90p 220s 44s/p/w

### Mac VirtualBox

#### 4 cores MacBook Pro 2.5g i7 16g DDR3

1c **3p** 41s 14s/p/w
3c 6p 42s 14s/p/w
4c 9p 49s 16s/p/w

**Comment by robnagler** [comment]
Ran on AWS, which shows linear behavior up to 16 cores, but then goes non-linear. I'm running more tests and will report here.

### iad1 c3.8xlarge E5-2680 v2 @ 2.80GHz 58gb

1c **3p** 71s 24s/p/w
5c 12p 72s 24s/p/w
9c 24p 73s 24s/p/w
16c 45p 78s 26s/p/w
18c 51p 100s 33s/p/w
20c 57p 102s 34s/p/w
22c 63p 107s 36s/p/w
24c 69p 115s 38s/p/w
26c 75p 122s 41s/p/w
28c 81p 124s 41s/p/w
30c 87p 135s 45s/p/w
32c 93p 147s 49s/p/w

1c **4p** 89s 22s/p/w
5c 16p 90s 23s/p/w
9c 32p 93s 23s/p/w
16c 60p 96s 24s/p/w
32c 124p 182s 46s/p/w

**Comment by robnagler** [comment]
There's a significant difference between c3 and c4 architecture, even though the price is lower for c4.8xlarge ($1.591/h) vs c3.8xlarge ($1.68/h).

After 16 cores, things get non-linear. It might be related to SRW.

### iad1 c3.8xlarge 32 x E5-2680 v2 @ 2.80GHz 58gb

1c **10p** 204s 20s/p/w
4c 30p 200s 20s/p/w
8c 70p 206s 21s/p/w
12c 110p 208s 21s/p/w
16c 150p 216s 22s/p/w
20c 190p 296s 30s/p/w
24c 230p 315s 32s/p/w
28c 270p 366s 37s/p/w
32c 310p 399s 40s/p/w

### iad2 c4.8xlarge 36 x E5-2666 v3 @ 2.90GHz 62gb

1c **5p** 77s 15s/p/w
5c 20p 77s 15s/p/w
9c 40p 80s 16s/p/w
12c 55p 81s 16s/p/w
16c 75p 83s 17s/p/w
20c 95p 112s 22s/p/w
24c 115p 121s 24s/p/w
28c 135p 128s 26s/p/w
32c 155p 140s 28s/p/w
36c 175p 162s 32s/p/w

**Comment by cchall** [comment]
Synergia timings on iad2:

With 102400 macroparticles and space charge:

Cores: 1   Time/Turn: 43.3210576
Cores: 2   Time/Turn: 22.474906
Cores: 3   Time/Turn: 15.6230822
Cores: 4   Time/Turn: 12.0003902
Cores: 8   Time/Turn: 6.5445916
Cores: 12   Time/Turn: 4.773387
Cores: 16   Time/Turn: 3.9343014
Cores: 20   Time/Turn: 5.3149844
Cores: 24   Time/Turn: 5.0195374
Cores: 28   Time/Turn: 4.7463082
Cores: 32   Time/Turn: 4.3150172
Cores: 36   Time/Turn: 4.2205376

Scaling macroparticle number with cores, starting at 25600 particles:

Cores: 1   Time/Turn: 11.3864818
Cores: 2   Time/Turn: 11.625689
Cores: 3   Time/Turn: 12.0406684
Cores: 4   Time/Turn: 11.9440856
Cores: 8   Time/Turn: 12.5271736
Cores: 12   Time/Turn: 12.599958
Cores: 16   Time/Turn: 13.05489
Cores: 20   Time/Turn: 22.904696
Cores: 24   Time/Turn: 23.6713924
Cores: 28   Time/Turn: 24.2263492
Cores: 32   Time/Turn: 25.3413938
Cores: 36   Time/Turn: 26.6965274

**Comment by robnagler** [comment]
Similar performance elbow at 16 cores. Now I'm really curious.

**Comment by robnagler** [comment]
Getting better at deploying. Took less than 5 minutes once the account was setup on SoftLayer to deploy sjc1. There's less of an elbow on this one.

### sjc1 32 x E5-2683 v3 @ 2.00GHz, 132GB

1c **5p** 131s 26s/p/w
5c 20p 124s 25s/p/w
9c 40p 127s 25s/p/w
14c 65p 131s 26s/p/w
15c 70p 131s 26s/p/w
16c 75p 139s 28s/p/w
17c 80p 135s 27s/p/w
18c 85p 142s 28s/p/w
20c 95p 137s 27s/p/w
24c 115p 152s 30s/p/w
26c 125p 177s 35s/p/w
28c 135p 185s 37s/p/w
31c 150p 189s 38s/p/w
32c 155p 192s 38s/p/w

**Comment by robnagler** [comment]
@cchall ran simultaneous synergia runs on iad2 with very little loss. I did the same for sjc1, iad2, and dfw13. iad2 did better than the others as the number of cores in use increased, but on the lower core runs, iad2 was less efficient than the non-concurrent runs with same cores.

sjc1 is a E5-2683 v3 with 14 cores per socket.  dfw13 is a E5-2680 v2 with 10 cores per socket. iad2 runs on a [secret AWS chip](https://www.theregister.co.uk/2015/01/13/amazons_new_ec2_compute_instances_run_on_secret_intel_chips/) called the E5-2666, which is not spec'd on intel.com, but it seems that it is an 18 core chip with 36 threads ("vCPUs" in AWS-speak). However, it seems to handle contention better than the other VMs, which justifies this quote from [the article](https://www.theregister.co.uk/2015/01/13/amazons_new_ec2_compute_instances_run_on_secret_intel_chips/): "AWS veep James Hamilton explained to El Reg at AWS re:Invent. 'As a consequence, the processors can be pushed harder. Through partnership with Intel, we have processors that are actually running faster at a given core count than is available on the open market.'"

It's all very interesting, and deserves study. I'm trying to get Google (GCP) to give us 64 cores. (Initial accounts have a max of 24 cores.)

### dfw13 20 core Linode

1c 5p 130s 26s/p/w
5c 20p 254s 51s/p/w
9c 40p 264s 53s/p/w
10c 45p 264s 53s/p/w

1c 5p 130s 26s/p/w
5c 20p 253s 51s/p/w
9c 40p 264s 53s/p/w
10c 45p 264s 53s/p/w

### sjc1 32 cores SoftLayer

1c 5p 129s 26s/p/w
5c 20p 247s 49s/p/w
9c 40p 258s 52s/p/w
15c 70p 263s 53s/p/w
16c 75p 269s 54s/p/w

1c 5p 129s 26s/p/w
5c 20p 247s 49s/p/w
9c 40p 258s 52s/p/w
15c 70p 262s 52s/p/w
16c 75p 271s 54s/p/w

### iad2 36 cores AWS

1c 5p 140s 28s/p/w
5c 20p 146s 29s/p/w
9c 40p 148s 30s/p/w
16c 75p 154s 31s/p/w
18c 85p 158s 32s/p/w

1c 5p 140s 28s/p/w
5c 20p 146s 29s/p/w
9c 40p 148s 30s/p/w
16c 75p 155s 31s/p/w
18c 85p 158s 32s/p/w

**Comment by bruhwile** [comment]
Thanks @cchall and @robnagler for the weak scaling data.

Assuming no underlying hardware complications, Amdahl's law is a useful guess for when/why the efficiency rolls over.  See e.g. https://en.wikipedia.org/wiki/Amdahl's_law There are more sophisticated formulations, but it's useful to just think about what percentage of the work has not been parallelized.

For example, the time MPI takes to instantiate itself and set up all the communicators, and then to instantiate the C++ executable on every processor, and then to instantiate Python, and then to import all the required libraries, etc. is part of the serial effort. File I/O tends to fall in this category, although parallel I/O obviously helps. Any MPI barriers, blocking sends or receives, or collective methods contribute to the serial fraction of the effort. Algorithms that rely heavily on processor 0 to do special tasks tend to drive up the serial fraction significantly.

Eyeballing the plots on the Amdahl wikipedia page indicates that 10% to 25% serial effort will cause the efficiency to roll over at 16 to 32 processors. One way to get some insight into this issue is to consider a test case where more work is done for each macro-particle and/or to increase the number of macro-particles per processor.

I'm not requesting additional runs -- just commenting.

**Comment by robnagler** [comment]
Amdahl's law applies. What isn't clear to me is why the the results are so different at times between providers. AWS allocates its vCPUs efficiently with two concurrent runs, whereas SoftLayer and Linode don't. Even more curious to me that it gets maximum efficiency out of its hyperthreading.

SoftLayer has a smoother elbow which starts later (24 cores vs 16). Yet, it thrashes with two concurrent simulations. Also, SoftLayer seems to have less consistency than other services. I did quite a few runs and the numbers move around a lot especially on the higher numbers of cores.

### sjc1 SoftLayer 32 core

1c 5p 129s 26s/p/w
2c 5p 120s 24s/p/w
3c 10p 123s 25s/p/w
4c 15p 124s 25s/p/w
5c 20p 124s 25s/p/w
6c 25p 125s 25s/p/w
7c 30p 127s 25s/p/w
8c 35p 129s 26s/p/w
9c 40p 127s 25s/p/w
10c 45p 128s 26s/p/w
11c 50p 131s 26s/p/w
12c 55p 130s 26s/p/w
13c 60p 131s 26s/p/w
14c 65p 131s 26s/p/w
15c 70p 132s 26s/p/w
16c 75p 134s 27s/p/w
17c 80p 136s 27s/p/w
18c 85p 137s 27s/p/w
19c 90p 137s 27s/p/w
20c 95p 141s 28s/p/w
21c 100p 156s 31s/p/w
22c 105p 148s 30s/p/w
23c 110p 151s 30s/p/w
24c 115p 149s 30s/p/w
25c 120p 172s 34s/p/w
26c 125p 168s 34s/p/w
27c 130p 159s 32s/p/w
28c 135p 149s 30s/p/w
29c 140p 197s 39s/p/w
30c 145p 180s 36s/p/w
31c 150p 206s 41s/p/w
32c 155p 195s 39s/p/w

1c 5p 130s 26s/p/w
2c 5p 120s 24s/p/w
3c 10p 121s 24s/p/w
4c 15p 123s 25s/p/w
5c 20p 124s 25s/p/w
6c 25p 125s 25s/p/w
7c 30p 127s 25s/p/w
8c 35p 129s 26s/p/w
9c 40p 130s 26s/p/w
10c 45p 130s 26s/p/w
11c 50p 131s 26s/p/w
12c 55p 132s 26s/p/w
13c 60p 131s 26s/p/w
14c 65p 132s 26s/p/w
15c 70p 137s 27s/p/w
16c 75p 135s 27s/p/w
17c 80p 137s 27s/p/w
18c 85p 139s 28s/p/w
19c 90p 146s 29s/p/w
20c 95p 151s 30s/p/w
21c 100p 142s 28s/p/w
22c 105p 147s 29s/p/w
23c 110p 147s 29s/p/w
24c 115p 145s 29s/p/w
25c 120p 148s 30s/p/w
26c 125p 163s 33s/p/w
27c 130p 158s 32s/p/w
28c 135p 165s 33s/p/w
29c 140p 169s 34s/p/w
30c 145p 191s 38s/p/w
31c 150p 198s 40s/p/w
32c 155p 198s 40s/p/w

1c 5p 129s 26s/p/w
2c 5p 121s 24s/p/w
3c 10p 121s 24s/p/w
4c 15p 123s 25s/p/w
5c 20p 123s 25s/p/w
6c 25p 124s 25s/p/w
7c 30p 129s 26s/p/w
8c 35p 127s 25s/p/w
9c 40p 128s 26s/p/w
10c 45p 129s 26s/p/w
11c 50p 131s 26s/p/w
12c 55p 130s 26s/p/w
13c 60p 131s 26s/p/w
14c 65p 133s 27s/p/w
15c 70p 132s 26s/p/w
16c 75p 134s 27s/p/w
17c 80p 142s 28s/p/w
18c 85p 139s 28s/p/w
19c 90p 142s 28s/p/w
20c 95p 139s 28s/p/w
21c 100p 144s 29s/p/w
22c 105p 143s 29s/p/w
23c 110p 144s 29s/p/w
24c 115p 143s 29s/p/w
25c 120p 179s 36s/p/w
26c 125p 190s 38s/p/w
27c 130p 176s 35s/p/w
28c 135p 170s 34s/p/w
29c 140p 186s 37s/p/w
30c 145p 189s 38s/p/w
30c 145p 187s 37s/p/w
31c 150p 192s 38s/p/w
32c 155p 197s 39s/p/w

**Comment by robnagler** [comment]
I've added Google Compute Platform (GCP) into the mix. It seems to have the best overall performance (see graph). We don't know what processors it's running, because it just lists "Xeon CPU @ 2.3GHz".

![VPS performance (SRW Ex 10) 3/20/2017](https://cloud.githubusercontent.com/assets/5495179/24115800/27b1ed42-0d6a-11e7-9d88-3b6f594b406e.png)
[Source: Google Spreadsheet](https://docs.google.com/spreadsheets/d/1vcRzB2eFuNsG3jWjr0wtF0rhMjTij6ohfJblSdTeryU/edit#gid=1147270072)

### dsm1 GCP Xeon 2.3ghz 128gb (two concurrent jobs)

GCP doesn't perform as well as AWS with two concurrent jobs.

8c 35p 96s 19s/p/w
16c 75p 104s 21s/p/w
20c 95p 121s 24s/p/w
24c 115p 152s 30s/p/w
28c 135p 165s 33s/p/w
32c 155p 189s 38s/p/w

8c 35p 94s 19s/p/w
16c 75p 97s 19s/p/w
20c 95p 119s 24s/p/w
24c 115p 141s 28s/p/w
28c 135p 178s 36s/p/w
32c 155p 191s 38s/p/w

**Comment by cchall** [comment]
Tried starting up 3 concurrent synergia runs of 16 cores each on dsm1. With
one run time/turn was ~4.5 s, with two concurrent time/turn ~6 s, with the
third going concurrently all three ground to a halt and nothing completed a
single turn after 3+ minutes. Killing the third job promptly restored
progress of the other two.

**Comment by robnagler** [comment]
FTR, running two concurrent jobs on an AWS c4.8xlarge with synergia and no space charge doesn't seem promising.

### c4.8xlarge parallel

1c 58s 58s/c
2c 107s 53s/c
3c 109s 36s/c
4c 117s 29s/c
5c 120s 24s/c
6c 125s 20s/c
7c 127s 18s/c
8c 134s 16s/c
9c 142s 15s/c
10c 145s 14s/c
11c 165s 15s/c
12c 175s 14s/c
13c 211s 16s/c
14c 223s 15s/c
15c 246s 16s/c
16c 259s 16s/c
17c 284s 16s/c
18c 294s 16s/c

1c 58s 58s/c
2c 107s 53s/c
3c 110s 36s/c
4c 116s 29s/c
5c 120s 24s/c
6c 124s 20s/c
7c 128s 18s/c
8c 134s 16s/c
9c 143s 15s/c
10c 144s 14s/c
11c 165s 15s/c
12c 175s 14s/c
13c 210s 16s/c
14c 223s 15s/c
15c 245s 16s/c
16c 259s 16s/c
17c 284s 16s/c
18c 294s 16s/c

**Comment by robnagler** [comment]
Full data for m620

### m620

1c 5p 104s 21s/p/w
2c 5p 101s 20s/p/w
3c 10p 101s 20s/p/w
4c 15p 103s 21s/p/w
5c 20p 105s 21s/p/w
6c 25p 107s 21s/p/w
7c 30p 112s 22s/p/w
8c 35p 113s 23s/p/w
9c 40p 117s 23s/p/w
10c 45p 116s 23s/p/w
11c 50p 119s 24s/p/w
12c 55p 117s 23s/p/w
13c 60p 117s 23s/p/w
14c 65p 117s 23s/p/w
15c 70p 119s 24s/p/w
16c 75p 119s 24s/p/w
17c 80p 129s 26s/p/w
18c 85p 133s 27s/p/w
19c 90p 137s 27s/p/w
20c 95p 144s 29s/p/w
21c 100p 153s 31s/p/w
22c 105p 157s 31s/p/w
23c 110p 162s 32s/p/w
24c 115p 168s 34s/p/w
25c 120p 171s 34s/p/w
26c 125p 181s 36s/p/w
27c 130p 183s 37s/p/w
28c 135p 191s 38s/p/w
29c 140p 195s 39s/p/w
30c 145p 202s 40s/p/w
31c 150p 207s 41s/p/w
32c 155p 221s 44s/p/w

**Comment by robnagler** [comment]
@cchall are you satisfied with your setup?

**Comment by cchall** [comment]
Yep, seems to be working well.

**Comment by robnagler** [comment]
@cchall iad2 is up. 

I also ran some tests:

### r4.4xlarge:  Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz, 122GB

1c 5p 90s 18s/p/w
2c 5p 83s 17s/p/w
3c 10p 84s 17s/p/w
4c 15p 84s 17s/p/w
5c 20p 84s 17s/p/w
6c 25p 85s 17s/p/w
7c 30p 89s 18s/p/w
8c 35p 87s 17s/p/w
9c 40p 117s 23s/p/w
10c 45p 119s 24s/p/w
11c 50p 129s 26s/p/w
12c 55p 124s 25s/p/w
13c 60p 137s 27s/p/w
14c 65p 147s 29s/p/w
15c 70p 150s 30s/p/w
16c 75p 160s 32s/p/w
